src/main/scala/viper/server/vsi/VerificationServerInterface.scala

```diff
@@ -0,0 +1,310 @@
+        bindingFuture match {
+          case Some(future) =>
+            future
+              .flatMap(_.unbind()) // trigger unbinidng from the port
+              .onComplete(_ => {
+                system.terminate() // and shutdown when done
+              })
+          case None =>
+            system.terminate() // shutdown
+        }
+      case Terminator.WatchJobQueue(jid, handle) =>
+        val queue_completion_future: Future[Done] = handle.queue.watchCompletion()
+        queue_completion_future.onComplete( {
+          case Failure(e) =>
+            println(s"Terminator detected failure in job #$jid: $e")
+            throw e
+          case Success(_) =>
+            jobs.discardJob(jid)
+            println(s"Terminator deleted job #$jid")
+        })
+    }
+  }
+
+
+  // --- Actor: JobActor ---
+
+  object JobActor {
+    def props(id: Int): Props = Props(new JobActor(id))
+  }
+
+  class JobActor(private val id: Int) extends Actor {
+
+    private var _verificationTask: Thread = _
+
+    private def interrupt: Boolean = {
+      if (_verificationTask != null && _verificationTask.isAlive) {
+        _verificationTask.interrupt()
+        _verificationTask.join()
+        println(s"Job #$id has been successfully interrupted.")
+        return true
+      }
+      false
+    }
+
+    private def resetVerificationTask(): Unit = {
+      if (_verificationTask != null && _verificationTask.isAlive) {
+        _verificationTask.interrupt()
+        _verificationTask.join()
+      }
+      _verificationTask = null
+    }
+
+    override def receive: PartialFunction[Any, Unit] = {
+      //Agnostic
+      case VerificationProtocol.Stop(call_back_needed) =>
+        val did_I_interrupt = interrupt
+        if (call_back_needed) {
+          // If a callback is expected, then the caller must decide when to kill the actor.
+          if (did_I_interrupt) {
+            sender ! s"Job #$id has been successfully interrupted."
+          } else {
+            sender ! s"Job #$id has already been finalized."
+          }
+        }
+      case VerificationProtocol.Verify(task, queue, publisher) =>
+        resetVerificationTask()
+        sender ! startJob(task, queue, publisher)
+      case msg =>
+        throw new Exception("Main Actor: unexpected message received: " + msg)
+    }
+
+    private def startJob(task: Thread, queue: SourceQueueWithComplete[Envelope], publisher: Publisher[Envelope]): JobHandle = {
+      _verificationTask = task
+      _verificationTask.start()
+      println(s"Starting job #$id...")
+      JobHandle(self, queue, publisher)
+    }
+  }
+
+
+  // --- Actor: MessageActor ---
+
+  object MessageActor {
+    def props(jid: Int, queue: SourceQueueWithComplete[Envelope]): Props = Props(new MessageActor(jid, queue))
+  }
+
+  class MessageActor(jid: Int, queue: SourceQueueWithComplete[Envelope]) extends Actor {
+
+    override def receive: PartialFunction[Any, Unit] = {
+      case TaskProtocol.ClientRequest =>
```

Comment:
Why do we need this?

Reply:
The Message actor is what was originally called ReporterActor and then QueueActor. I wan't happy with the name and changed it to MessageActor. Looking back I'm not sure I like it, though.

Anyway, it's purpose is to offer messages to the SourceQueue that is holds messages for a VerificationTask (typically generated by some backend).